From 2457c06e03b1a291433ac676e4f4639f33668f12 Mon Sep 17 00:00:00 2001
From: Weilong Chen <chenweilong@huawei.com>
Date: Tue, 11 Apr 2023 17:16:37 +0800
Subject: [PATCH 2/2] net: Provide the idents_entries parameter to control the
 number of IP idents entries

In a scenario with sufficient memory, IP idents occupy 2M memory,
but the SOC will not always face such a large network pressure.

Signed-off-by: Weilong Chen <chenweilong@huawei.com>
---
 net/ipv4/route.c | 20 ++++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/net/ipv4/route.c b/net/ipv4/route.c
index 216f76b54..356acdfc0 100644
--- a/net/ipv4/route.c
+++ b/net/ipv4/route.c
@@ -3588,15 +3588,31 @@ static __net_initdata struct pernet_operations ipv4_inetpeer_ops = {
 struct ip_rt_acct __percpu *ip_rt_acct __read_mostly;
 #endif /* CONFIG_IP_ROUTE_CLASSID */
 
+static __initdata unsigned long idents_entries;
+static int __init set_idents_entries(char *str)
+{
+	ssize_t ret;
+
+	if (!str)
+		return 0;
+
+	ret = kstrtoul(str, 0, &idents_entries);
+	if (ret)
+		return 0;
+
+	return 1;
+}
+__setup("idents_entries=", set_idents_entries);
+
 int __init ip_rt_init(void)
 {
 	void *idents_hash;
 	int cpu;
 
-	/* For modern hosts, this will use 2 MB of memory */
+	/* For modern hosts, this will use 2 MB of memory or use idents_entries */
 	idents_hash = alloc_large_system_hash("IP idents",
 					      sizeof(*ip_idents) + sizeof(*ip_tstamps),
-					      0,
+					      idents_entries,
 					      16, /* one bucket per 64 KB */
 					      HASH_ZERO,
 					      NULL,
-- 
2.34.1

